### PR TITLE
#43: Fallback Mechanism for Lang Models

### PR DESCRIPTION
**Pull request by roma-glushko**
PoC of fallback mechanism for language models & the priority routing mechanism.

## Changes
- Implemented lang router handling flow 
- Implemented the priority routing strategy
- Implemented exponential retry when no model is healthy
- Implemented model health (based on the token bucket) and basic rate limit tracker
- Implemented a custom error budget field to accept budgets in form of human-friendly string e.g. "10/m".
- Moved build/factory methods on the config level

### CODE REVIEW

File: pkg/routers/health/buckets.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+	// Take into account burst size.
+	if minTime > timePointer {
+		newTime = minTime
+	}
+
+	return float64(now-newTime) / float64(b.timePerToken)
+}
+
+func (b *TokenBucket) nowInMicro() uint64 {
```
Comment (mkrueger12) [inline pkg/routers/health/buckets.go:85]: What does this do?

File: pkg/routers/health/buckets.go
Comment (mkrueger12) [inline pkg/routers/health/buckets.go:1]: Excited to give this a go!

File: pkg/routers/health/buckets_test.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import (
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestTokenBucket_Take(t *testing.T) {
+	bucketSize := 10
```
Comment (mkrueger12) [inline pkg/routers/health/buckets_test.go:12]: How do we decide on bucket size?

File: pkg/routers/health/error_budget.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+		unit:   MIN,
+	}
+}
+
+// Budget defines max allows number of errors per given time period
+func (b *ErrorBudget) Budget() uint64 {
+	return uint64(b.budget)
+}
+
+// TimePerTokenMicro defines how much time do we need to wait to get one error token recovered (in microseconds)
```
Comment (mkrueger12) [inline pkg/routers/health/error_budget.go:46]: Is this adjustable?

File: pkg/routers/retry/exp_test.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+func TestExpRetry_WaitTime(t *testing.T) {
+	maxRetries := 4
+	maxDelay := 10 * time.Millisecond
+	expectedDelays := []time.Duration{
+		2 * time.Millisecond,
+		4 * time.Millisecond,
+		8 * time.Millisecond,
+		10 * time.Millisecond,
+		10 * time.Millisecond,
```
Comment (mkrueger12) [inline pkg/routers/retry/exp_test.go:37]: Are both of these supposed to be set to 10?

**Review by mkrueger12 — APPROVED** [review]

File: pkg/routers/health/error_budget.go (hunk 3 above)
Comment (roma-glushko) [reply pkg/routers/health/error_budget.go:46]: @mkrueger12 yes, it's adjustable and I even connect error budget field to the general config already:
- https://github.com/EinStack/glide/pull/68/files#diff-91b58caeed8701c64ca63ce432a7adfafd02fdc7aca313c113c08c15122cf666
- https://github.com/EinStack/glide/pull/68/files#diff-093f2a2075da7c408d072d5cfbad2cde9608bdc90652f35eaab0100be4cc50eaR20

Users will be able to set up error budget in a form of "10/s" e.g. "tolerate at most 10 errors per second"

File: pkg/routers/health/buckets_test.go (hunk 2 above)
Comment (roma-glushko) [reply pkg/routers/health/buckets_test.go:12]: Bucket size is essentially the number in the error budget config e.g. "10/s". So the bucket is how many errors we could tolerate during the give time frame.

File: pkg/routers/health/buckets.go (hunk 1 above)
Comment (roma-glushko) [reply pkg/routers/health/buckets.go:85]: Gets timestamp in microseconds (it's in nanoseconds by the default)

File: pkg/routers/health/buckets.go
Comment (roma-glushko) [reply pkg/routers/health/buckets.go:1]: Me too 🙌 🙌 🙌

**Pull request merged**